HeterogeneousCore/SonicCore/src/SonicClientBase.cc

```diff
@@ -40,19 +59,39 @@ void SonicClientBase::start(edm::WaitingTaskWithArenaHolder holder) {
   holder_ = std::move(holder);
 }
 
-void SonicClientBase::start() { tries_ = 0; }
+void SonicClientBase::start() { 
+    tries_ = 0; 
+    // initialize all actions
+    for (const auto& action : retryActions_) {
+        action->start();
+    }
+}
 
 void SonicClientBase::finish(bool success, std::exception_ptr eptr) {
   //retries are only allowed if no exception was raised
   if (!success and !eptr) {
-    ++tries_;
-    //if max retries has not been exceeded, call evaluate again
-    if (tries_ < allowedTries_) {
-      evaluate();
-      //avoid calling doneWaiting() twice
-      return;
+    //++tries_;
+    ////if max retries has not been exceeded, call evaluate again
+    //if (tries_ < allowedTries_) {
+    //  evaluate();
+    //  //avoid calling doneWaiting() twice
+    //  return;
+    //}
+
+    // Check if any retry actions are still valid
+    bool anyRetryAllowed = false;
+    for (const auto& action : retryActions_) {
+        if (action->shouldRetry()) {
+            action->retry();  // Call retry only if shouldRetry_ is true
+            return;
+        }
+    }
+    // If no actions allow retries, stop retrying
+    if (!anyRetryAllowed) {
+        edm::LogInfo("SonicClientBase") << "No retry actions available. Stopping retries.";
+        return;
     }
-    //prepare an exception if exceeded
+    //prepare an exception if no more retries left
```

Comment:
this should not be an `else` with respect to the previous condition? i.e. if it runs out of retries, it should issue the exception